File: tovtk.jl
using LightXML

function tovtk(VX,VY,EToV,vals,valsname)

    xdoc = XMLDocument();

    xroot = create_root(xdoc, "VTKFile");
    set_attribute(xroot,"type","UnstructuredGrid");
    set_attribute(xroot,"version","0.1");
    set_attribute(xroot,"byte_order","LittleEndian");
   #set_attribute(xroot,"compressor","vtkZLibDataCompressor");


    nPoints = length(VX);
    (K,nFaces) = size(EToV);

    unstructgrid = new_child(xroot, "UnstructuredGrid");

    piece        = new_child(unstructgrid,"Piece");
    set_attribute(piece,"NumberOfPoints",@sprintf("%d",nPoints));
    set_attribute(piece,"NumberOfCells",@sprintf("%d",K));

    pointdata = new_child(piece,"PointData");
    set_attribute(pointdata,"Scalars",valsname);

    valsarray = new_child(pointdata,"DataArray");
    set_attribute(valsarray,"type","Float32");
    set_attribute(valsarray,"Name",valsname);
    set_attribute(valsarray,"format","ascii");
    add_text(valsarray,@sprintf("\n"));
    for i = 1 : nPoints 
        #v1=EToV[k,1];v2=EToV[k,2];v3=EToV[k,3];
        add_text(valsarray,@sprintf("%f\n",vals[i]));
    end

    celldata = new_child(piece,"CellData");

    points = new_child(piece,"Points");
    pointsarray = new_child(points,"DataArray");
    set_attribute(pointsarray,"type","Float32");
    set_attribute(pointsarray,"format","ascii");
    set_attribute(pointsarray,"NumberOfComponents","3");
    add_text(pointsarray,@sprintf("\n"));
    for i = 1 : nPoints
        add_text(pointsarray,@sprintf("%f %f %f\n",VX[i],VY[i],0.0));
    end

    cells = new_child(piece,"Cells");

    cellarray = new_child(cells,"DataArray");
    set_attribute(cellarray,"type","Int32");
    set_attribute(cellarray,"Name","connectivity");
    set_attribute(cellarray,"format","ascii");
    add_text(cellarray,@sprintf("\n"));
    for k = 1 : K
        add_text(cellarray,@sprintf("%d %d %d\n",
        EToV[k,1]-1,EToV[k,2]-1,EToV[k,3]-1));
    end

    offs = new_child(cells,"DataArray");
    set_attribute(offs,"type","Int32");
    set_attribute(offs,"Name","offsets");
    set_attribute(offs,"format","ascii");
    offset=3;
    add_text(offs,@sprintf("\n"));
    for k = 1 : K
        add_text(offs,@sprintf("%d ",offset));
        offset=offset+3;
    end


    celltypes = new_child(cells,"DataArray");
    set_attribute(celltypes,"type","UInt8");
    set_attribute(celltypes,"Name","types");
    set_attribute(celltypes,"format","ascii");
    offset=3;
    add_text(celltypes,@sprintf("\n"));
    for k = 1 : K
        add_text(celltypes,@sprintf("%d ",5));
    end




 




    save_file(xdoc, "test.vtu");


end

File: src/tovtk.c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <libxml/parser.h>
#include <libxml/tree.h>
#include "array.h"
#include "types.h"
#if defined(LIBXML_TREE_ENABLED) && defined(LIBXML_OUTPUT_ENABLED)

static void xmlAddText(xmlNodePtr node,const char* txt); 

void tovtk(oned_array* scalar_field, const char* fieldname,
		oned_array* vx, oned_array* vy, oned_array* vz,
		twod_array* eltovert,
		const char* filename){

	assert(vx->get_len(vx) == vy->get_len(vy));


	char buf[256];

	xmlDocPtr doc = NULL;
	xmlNodePtr root_node = NULL, unstructuredgrid = NULL, piece = NULL, pointdata = NULL, valsarray = NULL;
	xmlNodePtr celldata = NULL, points = NULL, pointsarray = NULL, cells = NULL, cellarray = NULL, offs = NULL;
	xmlNodePtr celltypes = NULL;

	doc = xmlNewDoc(BAD_CAST "1.0");
	root_node = xmlNewNode(NULL, BAD_CAST "VTKFile");
	xmlNewProp(root_node, BAD_CAST "type", BAD_CAST "UnstructuredGrid");
	xmlNewProp(root_node, BAD_CAST "version",BAD_CAST "0.1");
	xmlNewProp(root_node, BAD_CAST "byte_order", BAD_CAST "LittleEndian");
	xmlDocSetRootElement(doc,root_node);

	long npoints = vx->get_len(vx);
	long nelements = eltovert->get_rows(eltovert);
	long nvert     = eltovert->get_cols(eltovert);

	unstructuredgrid = xmlNewChild(root_node, NULL, BAD_CAST "UnstructuredGrid",NULL);
	piece = xmlNewChild(unstructuredgrid,NULL, BAD_CAST "Piece",NULL);

	sprintf(buf,"%ld",npoints);
	xmlNewProp(piece, BAD_CAST "NumberOfPoints",BAD_CAST buf);

	sprintf(buf,"%ld",npoints);
	xmlNewProp(piece,BAD_CAST "NumberOfCells",BAD_CAST buf);

	pointdata = xmlNewChild(piece,NULL, BAD_CAST "PointData",NULL);
	xmlNewProp(pointdata,BAD_CAST "Scalars",fieldname);


	valsarray = xmlNewChild(pointdata,NULL,BAD_CAST "DataArray",NULL);
	xmlNewProp(valsarray,BAD_CAST "type",BAD_CAST "Float32");
	xmlNewProp(valsarray,BAD_CAST "Name",BAD_CAST fieldname);
	xmlNewProp(valsarray,BAD_CAST "format",BAD_CAST "ascii");
	xmlAddText(valsarray,"\n");
	Index i;
	for(i=0;i<npoints;i++){
		Real val = scalar_field->get_val(scalar_field,i);
		sprintf(buf,"%f\n",val);
		xmlAddText(valsarray,BAD_CAST buf);
	}

	celldata = xmlNewChild(piece,NULL,BAD_CAST "CellData",NULL);
	points = xmlNewChild(piece,NULL,BAD_CAST "Points",NULL);
	pointsarray = xmlNewChild(points,NULL,BAD_CAST "DataArray",NULL);
	xmlNewProp(pointsarray,BAD_CAST "type",BAD_CAST "Float32");
	xmlNewProp(pointsarray,BAD_CAST "format",BAD_CAST "ascii");
	xmlNewProp(pointsarray,BAD_CAST "NumberOfComponents",BAD_CAST "3");
	xmlAddText(pointsarray, BAD_CAST "\n");
	for(i=0;i<npoints;i++){
		Real x = vx->get_val(vx,i);
		Real y = vy->get_val(vy,i);
		Real z = vz->get_val(vz,i);
		sprintf(buf,"%f %f %f\n",x,y,z);
		xmlAddText(pointsarray,BAD_CAST buf);
	}

	cells = xmlNewChild(piece,NULL,BAD_CAST "Cells",NULL);
	cellarray = xmlNewChild(cells,NULL,BAD_CAST "DataArray",NULL);
	xmlNewProp(cellarray,BAD_CAST "type", BAD_CAST "Int32");
	xmlNewProp(cellarray,BAD_CAST "Name",BAD_CAST "connectivity");
	xmlNewProp(cellarray,BAD_CAST "format",BAD_CAST "ascii");

	xmlAddText(cellarray,"\n");
	Index k,v;
	for(k=0;k<nelements;k++){
		for(v=0;v<nvert;v++){
			Index v1 = eltovert->get_val(eltovert,k,v);
			sprintf(buf,"%ld ",v1);
			xmlAddText(cellarray,BAD_CAST buf);
		}
		xmlAddText(cellarray,"\n");
	}

	offs = xmlNewChild(cells,NULL,BAD_CAST "DataArray",NULL);
	xmlNewProp(offs,BAD_CAST "type",BAD_CAST "Int32");
	xmlNewProp(offs,BAD_CAST "Name",BAD_CAST "offsets");
	xmlNewProp(offs,BAD_CAST "format",BAD_CAST "ascii");
	Index offset=nvert;
	for(k=0;k<nelements;k++){
		sprintf(buf,"%ld ",offset);
		xmlAddText(offs,BAD_CAST buf);
		offset=offset+nvert;
	}

	Index eltype;
	if(nvert==3)
		eltype=5;
	else
		eltype=10;

	celltypes = xmlNewChild(cells,NULL,BAD_CAST "DataArray",NULL);
	xmlNewProp(celltypes,BAD_CAST "type",BAD_CAST "UInt8");
	xmlNewProp(celltypes,BAD_CAST "Name",BAD_CAST "types");
	xmlNewProp(celltypes,BAD_CAST "format",BAD_CAST "ascii");
	xmlAddText(celltypes,"\n");
	sprintf(buf,"%ld ",eltype);
	for(k=0;k<nelements;k++){
		xmlAddText(celltypes,BAD_CAST buf);
	}

	xmlSaveFormatFileEnc(BAD_CAST filename,doc,BAD_CAST "UTF-8", 1);
	xmlFreeDoc(doc);
	xmlCleanupParser();


}


static void xmlAddText(xmlNodePtr node,const char* txt){
	xmlAddChild(node,xmlNewText(BAD_CAST txt));
}
#else
#warning libxml tree support not compiled.
#endif
